### PR TITLE
Feature ggomnbd

### PR DESCRIPTION
**Pull request by niels89**

### CODE REVIEW

**Comment by niels89** [comment]
@bachmannpatrick I have added my changes to the GGompertz/NBD model. After you have merged your documentation changes, I will merge development once again in order to account for the changes.

File: R/class_clv_model_ggomnbd_staticcov.R [hunk 1]
```diff
@@ -0,0 +1,201 @@
+#' CLV Model functionality for GGompertz/NBD with static covariates
+#'
+#' This class implements the functionalities and model-specific steps which are required
+#' to fit the GGompertz/NBD model with static covariates.
+#'
```
Comment (pschil) [inline R/class_clv_model_ggomnbd_staticcov.R:5]: template for these silly model docus

File: R/class_clv_model_ggomnbd_staticcov.R [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  # no transformation needed
+  return(prefixed.params.cov)
+})
+
+setMethod(f = "clv.model.put.newdata", signature = signature(clv.model = "clv.model.ggomnbd.static.cov"), definition = function(clv.model, clv.fitted, verbose){
+  # clv.data in clv.fitted is already replaced with newdata here
+  # Need to only redo cbs if given new data
+  clv.fitted@cbs <- ggomnbd_cbs(clv.data = clv.fitted@clv.data)
+  return(clv.fitted)
+})
```
Comment (pschil) [inline R/class_clv_model_ggomnbd_staticcov.R:69]: dont overwrite, use nocov

File: R/class_clv_model_ggomnbd_staticcov.R [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+setMethod(f = "clv.model.prepare.optimx.args", signature = signature(clv.model="clv.model.ggomnbd.static.cov"), definition = function(clv.model, clv.fitted, prepared.optimx.args,...){
+  # Do not call the no.cov function as the LL is different
+
+  # Everything to call the LL function
+  optimx.args <- modifyList(prepared.optimx.args,
```
Comment (pschil) [inline R/class_clv_model_ggomnbd_staticcov.R:75]: ensure/verify sorting of cbs vs cov data!

File: R/class_clv_model_ggomnbd_staticcov.R [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+#' @importFrom stats integrate
+setMethod("clv.model.expectation", signature(clv.model="clv.model.ggomnbd.static.cov"), function(clv.model, clv.fitted, dt.expectation.seq, verbose){
+  r <- alpha_i <- beta_i <- b <- s <- t_i <- tau <- NULL
+
+  params_i <- clv.fitted@cbs[, c("Id", "T.cal", "date.first.actual.trans")]
+  params_i[, r := clv.fitted@prediction.params.model[["r"]]]
+  params_i[, alpha_i := clv.fitted@prediction.params.model[["alpha"]] * exp( -clv.data.get.matrix.data.cov.trans(clv.fitted@clv.data) %*% clv.fitted@prediction.params.trans)]
+  params_i[, beta_i  := clv.fitted@prediction.params.model[["beta"]]  * exp( -clv.data.get.matrix.data.cov.life(clv.fitted@clv.data)  %*% clv.fitted@prediction.params.life)]
+  params_i[, b := clv.fitted@prediction.params.model[["b"]]]
+  params_i[, s := clv.fitted@prediction.params.model[["s"]]]
```
Comment (pschil) [inline R/class_clv_model_ggomnbd_staticcov.R:104]: indention + math

File: R/class_clv_model_ggomnbd_staticcov.R [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  r <- alpha <- beta <- b <- s <- CET <- PAlive <- DERT <- period.length <- NULL
+  # Covariates as matrix, if there is a covariate
+  data.cov.mat.life  <- clv.data.get.matrix.data.cov.life(clv.fitted@clv.data)
+  data.cov.mat.trans <- clv.data.get.matrix.data.cov.trans(clv.fitted@clv.data)
+
+
+  # To be sure they are both sorted the same when calling cpp functions
+  setkeyv(dt.prediction, "Id")
+  setkeyv(clv.fitted@cbs, "Id")
+
```
Comment (pschil) [inline R/class_clv_model_ggomnbd_staticcov.R:135]: ensure/verify sorting of cov data vs cbs

File: R/f_interface_ggomnbd.R [hunk 6]
```diff
@@ -1,52 +1,126 @@
...
-
-
+#' @template template_details_paramsggomnbd
+#'
+#' @details If no start parameters are given, r = 1, alpha = 1, beta = 1, b = 1, s = 1 is used.
+#' The model start parameters are required to be > 0.
+#'
+#' \subsection{The GGompertz / NBD model}{
+#' TODO
+#' }
```
Comment (pschil) [inline R/f_interface_ggomnbd.R:26]: @bachmannpatrick needs minimal explanations about the GGompertzNBD model

File: R/f_interface_ggomnbd.R [hunk 7]
```diff
@@ -1,52 +1,126 @@
...
+#' ggomnbd(clv.data.cdnow,
+#'      start.params.model = c(r=1, alpha=1, beta=1, b=1, s=1))
+#'
+#'
+#' # pass additional parameters to the optimizer (optimx)
+#' #    Use Nelder-Mead as optimization method and print
+#' #    detailed information about the optimization process
+#' estimation.ggomnbd <- ggomnbd(clv.data = clv.data.cdnow,
+#'                                   start.params.model = c(r = 1, alpha = 1, beta = 1, b = 1, s = 1))
+#'
```
Comment (pschil) [inline R/f_interface_ggomnbd.R:60]: does not do as commented

File: R/f_interface_ggomnbd.R [hunk 8]
```diff
@@ -1,52 +1,126 @@
...
+                                                                                       start.params.model=c(),
+                                                                                       use.cor = FALSE,
+                                                                                       start.param.cor=c(),
+                                                                                       optimx.args=list(),
+                                                                                       verbose=TRUE,...){
+  cl        <- match.call(call = sys.call(-1), expand.dots = TRUE)
+
+  obj <- clv.ggomnbd(cl=cl, clv.data=clv.data)
+
+  return(clv.template.controlflow.estimate(clv.fitted=obj, cl=cl, start.params.model = start.params.model, use.cor = use.cor,
```
Comment (pschil) [inline R/f_interface_ggomnbd.R:84]: block if use.cor=T

File: man-roxygen/template_details_paramsggomnbd.R [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#'
+#' @details
+#' Model parameters for the GGompertz /NBD model are \code{r, alpha, beta, b and s}. \cr
+#' \code{r}: TODO \cr
+#' \code{alpha}: TODO \cr
+#' \code{beta}: TODO \cr
+#' \code{b}: TODO \cr
+#' \code{s}: TODO
```
Comment (pschil) [inline man-roxygen/template_details_paramsggomnbd.R:8]: @bachmannpatrick some explanations about the coefs are needed here

File: tests/testthat/helper_s3_fitted.R [hunk 10]
```diff
@@ -21,12 +21,18 @@
     expect_named(res.coef, rownames(coef(summary(clv.fitted))), ignore.case = FALSE, ignore.order = FALSE)
   })
 
+  test_that("No cor if complete = FALSE", {
+    expect_false(clv.fitted@name.correlation.cor %in% names(coef(clv.fitted, complete=FALSE)))
+  })
+
```
Comment (pschil) [inline tests/testthat/helper_s3_fitted.R:27]: wasnt complete removed from coef/vcov previously?

File: tests/testthat/test_consistency_ggomnbd.R [hunk 11]
```diff
@@ -0,0 +1,55 @@
+skip_on_cran()
+
+# Consistency
+# nocov vs static cov:
+#   same fit with all covs = 0
+#   same predict with gamma=0
+
+context("Nocov/cov Consistency - BG/NBD - all cov data = 0")
```
Comment (pschil) [inline tests/testthat/test_consistency_ggomnbd.R:8]: GGompnbd

File: tests/testthat/test_consistency_ggomnbd.R [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+expect_silent(apparelStaticCov.0[1, Gender := 1])
+expect_silent(apparelStaticCov.0[, Channel := 0])
+expect_silent(apparelStaticCov.0[1, Channel := 1])
+expect_silent(clv.apparel.static <- SetStaticCovariates(clv.apparel,
+                                                        data.cov.life = apparelStaticCov.0, data.cov.trans = apparelStaticCov.0,
+                                                        names.cov.life = c("Gender", "Channel"), names.cov.trans = c("Gender", "Channel")))
+
+
+# Fit models
+expect_silent(g.nocov  <- ggomnbd(clv.apparel, verbose = FALSE))
```
Comment (pschil) [inline tests/testthat/test_consistency_ggomnbd.R:26]: Compare values of LL directly

File: tests/testthat/test_runability_ggomnbd_staticcov.R [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+# Generate fake cov data
+expect_silent(dt.covs <- data.table::rbindlist(lapply(paste0(LETTERS,1:100,sep=""), function(cid){
+  data.table::data.table(cid, Gender = sample(0:1, size = 1), Channel =  sample(0:1, size = 1))
+})))
+
+# create newdata objects with covariates
+expect_silent(clv.newdata.nohold <- SetStaticCovariates(
+  clv.data = clvdata(data.transactions = dt.newdata, date.format = "ydm", time.unit = "w",
+                     estimation.split = NULL, name.id = "cust.id", name.date = "trans.date",
```
Comment (pschil) [inline tests/testthat/test_runability_ggomnbd_staticcov.R:38]: this is copy-paste everywhere. move to function

File: R/f_interface_ggomnbd.R (hunk 7 above)
Comment (pschil) [reply R/f_interface_ggomnbd.R:60]: make template

**Pull request merged**